src/Voron/Impl/PagerState.cs

```diff
@@ -100,24 +100,75 @@ public PagerState(AbstractPager pager, long prefetchSegmentSize, long prefetchRe
 
         public readonly byte* MapBase; // { get; set; }
 
-        internal bool _released;
-
         public void Release()
         {
-            if (Interlocked.Decrement(ref _refs) != 0)
+            int currentRefCount;
+            int newRefCount;
+
+            // Uses atomic compare-and-swap to safely decrement the reference count,
+            // ensuring consistency even with concurrent modifications by other threads.
+            do
+            {
+                currentRefCount = _refs;
+
+                // Checks for negative reference counts to detect potential misuse or logical errors,
+                // such as double releases or illegal state transitions, preventing further incorrect decrements.
+                if (currentRefCount < 0)
+                    ThrowInvalidPagerState();
+
+                newRefCount = currentRefCount - 1;
+            }
+            while (Interlocked.CompareExchange(ref _refs, newRefCount, currentRefCount) != currentRefCount);
+
+            // If the reference count is still positive, there are active references.
+            if (newRefCount > 0)
                 return;
 
+            // Ensures reference count does not go negative, catching and preventing logical errors in reference management,
+            // thereby maintaining integrity in debug builds.
+            EnsureNoNegativeReferenceCount(newRefCount);
 
 #if DEBUG_PAGER_STATE
+            // In debug mode, removes the instance from tracking to help diagnose reference issues,
+            // ensuring the debug state is consistent and does not grow unbounded.
+
             String value;
             Instances.TryRemove(this, out value);
 #endif
 
             ReleaseInternal();
         }
 
+        [Conditional("DEBUG")]
+        [Conditional("VALIDATE")]
+        private void EnsureNoNegativeReferenceCount(int refCount)
+        {
+            // Validates that the reference count does not drop below zero,
+            // catching potential logical errors and preventing them from affecting reference management.
+            if (refCount < 0)
+                throw new InvalidOperationException("The ref count is negative. This can't happen");
+        }
+
+        public void EnsurePageStateIsValid()
+        {
+            if (_refs <= 0)
+                ThrowInvalidPagerState();
+        }
+
+        public bool IsReleased
+        {
+            get { return _refs < 0; }
```

Comment:
This is not consistent in terms of proper reference count conditions.

Reply:
It can still be valid but a reference was never acquired yet (not initialized). There are 2 options, either we allow that special case or we look into the caller code that we never check that if the page state has not been initialized yet. This is consistent in how it was working before. 

https://github.com/ravendb/ravendb/pull/18645/files#diff-3dc66b8a1b3ccfc5f1a4a04f771559ab49a7de68568b1bdf26be6204e5d08681R327